engines/mus.cpp

```diff
@@ -146,25 +150,48 @@ namespace pono {
 
     /*
      * If the conjunct encodes a state update constraint (i.e. of form
-     * <stateVar>.next = ...), then uses that state variable's id for the
+     * <stateVar>.next = ...), then use that state variable's id for the
      * control var symbol. Otherwise use the conjunct's to_string as an
      * identifier.
      */
-    UnorderedTermMap transIdToConjunct;
+    unordered_map<string, Term> transIdToConjunct;
     for (auto &tc: transConjuncts) {
       Term id = tc;
       if (tc->get_op() == PrimOp::Equal) {
         Term lhs = *tc->begin();
         if (ts_.is_next_var(lhs)) {
-          id = lhs;
+          id = ts_.curr(lhs);
         }
       }
       Term t = unrollUntilBound(tc, k);
       if (!options_.mus_include_yosys_internal_netnames_ && isYosysInternalNetname(id)) {
         solver_->assert_formula(t);
         assertions.push_back(t);
       } else {
-        transIdToConjunct.insert({id, t});
+        transIdToConjunct.insert({id->to_string(), t});
+      }
+    }
+
+    if (!options_.mus_combine_suffix_.empty()) {
+      /*
+       * Conjoin TRANS constraints that are identical up to suffix matching the
+       * supplied regular expression.
+       */
+      std::multimap<string, Term> mm;
+      for (auto & [id, tc] : unordered_map(transIdToConjunct)) {
+        smatch m;
+        regex r(std::regex("(.*)" + options_.mus_combine_suffix_));
+        if (regex_search(id, m, r)) {
+          mm.insert({m[1], tc});
+          transIdToConjunct.erase(id);
+        }
+      }
+      for (auto & [id, tc] : mm) {
+        if (transIdToConjunct.find(id) == transIdToConjunct.end()) {
+          transIdToConjunct[id] = tc;
```

Comment:
Checking my understanding:

If I have `"a_0" |-> t1` and `"a_1" |-> t2` in `transIdToConjunct` then I'd expect `mm` to have `"a" |-> [t1, t2]` and so the final value of `transIdToConjunct` should be `"a" |-> t2` (assuming `t1 < t2`)?

Reply:
Not exactly. You've got the first step right:

> If I have "a_0" |-> t1 and "a_1" |-> t2 in transIdToConjunct then I'd expect mm to have "a" |-> [t1, t2]

But when iterating over `mm` on line 190, there will be two passes through the loop for the `"a" |-> [t1, t2]` entry, for each of `("a", t1)` and `("a", t2)`. Say `("a", t1)` is processed first: `transIdToConjuct` doesn't have an entry for `"a"`, so `"a" |-> t1` is added (line 190-191). When `("a", t2)` is subsequently processed, it'll hit the `else` on line 192 and the entry for `"a"` in `transIdToConjunct` will be updated to `"a" |-> t1 /\ t2`

Reply:
Gotcha. I understand now. Thanks!

I think as I was executing this mentally, I interpreted `transIdToConjunct.find(id) == transIdToConjunct.end()` correctly on the first iteration. Then, my brain had a fault on the second iteration, and I interpreted it as "the element was found."

If only C++ had a `transIdToConjunct.contains(id)` or similar :(